src/theory/strings/theory_strings_rewriter.cpp

```diff
@@ -1775,55 +1760,116 @@ Node TheoryStringsRewriter::rewriteIndexof( Node node ) {
-          if( node[0].isConst() ){
-            return node[0];
-          }else{
-            //check for overlap, if none, we can remove the prefix
-            if( s.overlap(t)==0 ){
-              std::vector< Node > spl;
-              spl.insert( spl.end(), children.begin()+1, children.end() );
-              return NodeManager::currentNM()->mkNode( kind::STRING_CONCAT, children[0],
-                          NodeManager::currentNM()->mkNode( kind::STRING_STRREPL, mkConcat( kind::STRING_CONCAT, spl ), node[1], node[2] ) );
-            }
-          }
+      else
+      {
+        CVC4::String s1 = s.substr(0, (int)p);
+        CVC4::String s3 = s.substr((int)p + (int)t.size());
+        Node ns1 = NodeManager::currentNM()->mkConst(::CVC4::String(s1));
+        Node ns3 = NodeManager::currentNM()->mkConst(::CVC4::String(s3));
+        Node ret = NodeManager::currentNM()->mkNode(
+            kind::STRING_CONCAT, ns1, node[2], ns3);
+        return returnRewrite(node, ret, "rpl-const-find");
+      }
+    }
+  }
+
+  std::vector<Node> children0;
+  getConcat(node[0], children0);
+  std::vector<Node> children1;
+  getConcat(node[1], children1);
+
+  // check if contains definitely does (or does not) hold
+  Node cmp_con =
+      NodeManager::currentNM()->mkNode(kind::STRING_STRCTN, node[0], node[1]);
+  Node cmp_conr = Rewriter::rewrite(cmp_con);
+  if (cmp_conr.isConst())
+  {
+    if (cmp_conr.getConst<bool>())
+    {
+      // component-wise containment
+      std::vector<Node> cb;
+      std::vector<Node> ce;
+      int cc = componentContains(children0, children1, cb, ce, true, 1);
+      if (cc != -1)
+      {
+        if (cc == 0 && children0[0] == children1[0])
+        {
+          // definitely a prefix, can do the replace
+          std::vector<Node> cres;
+          cres.push_back(node[2]);
+          cres.insert(cres.end(), ce.begin(), ce.end());
+          Node ret = mkConcat(kind::STRING_CONCAT, cres);
+          return returnRewrite(node, ret, "rpl-cctn-rpl");
+        }
+        else if (ce.empty())
+        {
+          // we can pull remainder past first definite containment
+          std::vector<Node> cc;
+          cc.push_back(NodeManager::currentNM()->mkNode(
+              kind::STRING_STRREPL,
+              mkConcat(kind::STRING_CONCAT, children0),
+              node[1],
+              node[2]));
+          cc.insert(cc.end(), ce.begin(), ce.end());
+          Node ret = mkConcat(kind::STRING_CONCAT, cc);
+          return returnRewrite(node, ret, "rpl-cctn");
         }
       }
     }
+    else
+    {
+      // ~contains( t, s ) => ( replace( t, s, r ) ----> t )
+      return returnRewrite(node, node[0], "rpl-nctn");
+    }
   }
+  else if (cmp_conr != cmp_con)
+  {
+    // pull endpoints that can be stripped
+    std::vector<Node> cb;
+    std::vector<Node> ce;
+    if (stripConstantEndpoints(children0, children1, cb, ce))
+    {
+      std::vector<Node> cc;
+      cc.insert(cc.end(), cb.begin(), cb.end());
```

Comment:
Nit: the above to lines can be combined into: `std::vector<Node> cc(cb);`

Reply:
I think I prefer it this way since its symmetrical to the addition of ce at the end (makes it clear that we are miniscoping w.r.t. cb/ce).

Reply:
Yeah, that makes sense.